### ISSUE TITLE
Figure out why turns take longer as the game goes on, and fix it

### ISSUE THREAD

**Issue by QuintillusCFC** [C7-Game/Prototype#273]
Flintlock mentioned semi-recently that after about turn 20 things slowed down, and it sure does.  I've started investigating this, and found that galleys are always slow, and they're slow due to trying to find a path to their exploration destination.  Probably related to this, they can't move currently.  But sometimes other units are slow, too.  Log sample:

```
-> Begin Rome turn
Unit processing time for Chariot: 0 ms
Pre found new path time = 0 ms
Found new path time = 633 ms
Moved to new path time = 633 ms
Unit processing time for Galley: 633 ms
-> Begin Babylon turn
Unit processing time for Chariot: 0 ms
Set NEAR_CITIES exploration AI for Catapult with 1 movement points and 3 hit points, guid = 7130dd5f-c053-468f
-8257-2241ac68574b
Unit processing time for Catapult: 245 ms
Unit processing time for Warrior: 0 ms
Unit processing time for Chariot: 0 ms
Unit processing time for Warrior: 0 ms
Unit processing time for Chariot: 0 ms
Unit processing time for Warrior: 230 ms
Unit processing time for Chariot: 0 ms

```

I've added some logging on the SlowTurnsInvestigation branch to produce this, using the Stopwatch class for the timings.  200+ ms for one unit is obviously too slow.

This is tested by fortifying all my units at game start, and then pressing down both the Enter and Shift keys at the same time.  Enter keeps turns rolling, Shift skips all the animations.  So the only thing it's doing is processing the mechanics/AI.

The catapult happens to be bombarding a city in this log, so it's possible that some of the slowness is when combat occurs (even if animations are shift-skipped).  But the galleys are definitely pathfinding-related or pathfinding-adjacent.

(This is also an area where nifty logs might be useful.  For example, if we detect a unit taking longer than, say, 100ms, we should increase logging so we can investigate why (although this depends on having useful logging in the first place.  But at least a warning would be nice regardless).  Although we'd have to stop the timer for combat animations)

**Comment by QuintillusCFC** [comment]
More info.  Rome is at 67, 35, and built a galley.  With more logs, it prints:

```
--> Begin Rome turn
Unit processing time for Warrior: 0 ms
Unit processing time for Chariot: 0 ms
Pre found new path time = 0 ms
Valid explorer tile time = 0 ms
Pre algo time = 0 ms
Algo time for [67, 31] (Coast on Coast) = 16 ms
Algo time for [66, 32] (Coast on Coast) = 33 ms
Algo time for [63, 35] (Forest on Grassland) = 33 ms
Algo time for [68, 30] (Coast on Coast) = 49 ms
Algo time for [71, 33] (Coast on Coast) = 65 ms
Algo time for [69, 29] (Coast on Coast) = 81 ms
Algo time for [70, 30] (Coast on Coast) = 99 ms
Algo time for [71, 31] (Grassland on Grassland) = 99 ms
Algo time for [72, 32] (Coast on Coast) = 117 ms
Algo time for [72, 34] (Coast on Coast) = 134 ms
Algo time for [73, 35] (Coast on Coast) = 152 ms
Algo time for [72, 36] (Coast on Coast) = 169 ms
Algo time for [65, 31] (Coast on Coast) = 187 ms
Algo time for [63, 33] (Coast on Coast) = 205 ms
Algo time for [64, 32] (Coast on Coast) = 223 ms
Algo time for [62, 34] (Coast on Coast) = 240 ms
Algo time for [62, 36] (Coast on Coast) = 257 ms
Algo time for [64, 40] (Forest on Grassland) = 258 ms
Algo time for [69, 41] (Jungle on Grassland) = 259 ms
Algo time for [68, 42] (Grassland on Grassland) = 259 ms
Algo time for [67, 43] (Hills on Grassland) = 260 ms
Algo time for [66, 42] (Forest on Grassland) = 260 ms
Algo time for [65, 41] (Forest on Grassland) = 261 ms
Algo time for [61, 39] (Coast on Coast) = 278 ms
Algo time for [60, 38] (Coast on Coast) = 294 ms
Algo time for [61, 37] (Coast on Coast) = 312 ms
Algo time for [63, 41] (Grassland on Grassland) = 312 ms
Algo time for [62, 42] (Grassland on Grassland) = 313 ms
Algo time for [61, 41] (Coast on Coast) = 330 ms
Algo time for [60, 40] (Coast on Coast) = 346 ms
Algo time for [72, 38] (Coast on Coast) = 363 ms
Algo time for [73, 37] (Coast on Coast) = 380 ms
Algo time for [73, 39] (Coast on Coast) = 397 ms
Algo time for [72, 40] (Coast on Coast) = 414 ms
Algo time for [71, 41] (Jungle on Grassland) = 415 ms
Algo time for [70, 42] (Jungle on Grassland) = 416 ms
Post algo time = 416 ms
Debugging hopefully is useful
```

So it's doing pathing for 36 tiles, about 11.5 milliseconds per tile, though with some considerable variation.

It's noteworthy that all the tiles that take > 1 ms are coast.  Coast takes 16-18 ms typically.

I highly suspect this is due to our DijkstrasLandAlgorithm not being able to find a path to a sea tile, but trying everything in the book in the attempt.  But I'm a bit too tired to investigate that definitively right now.

It's also worth noting that each of those "algo time" statements is re-computing using Dijkstra from the same starting tile.  We knew this could be a hazard, as seem in the comment:

```
				// TODO: this is fastest if recomputing Dijkstra's for each unit
				// and ignoring that units may path from the same starting tile
				if (closest.Key == destination) {
					break;
				}
```

At a higher level, the reason the ExplorerAI is calling pathing from the same starting tile so many times is that it's trying to figure out what the shortest path is to an area where exploration is needed, and to do that, it's looking at all possible places to explore, and asking, "how far away is this tile from where this unit is located?"

Now in this particular case I don't think the "not caching if it's from the same start tile" case is making things worse for galleys, since we're asking for impossible routes, and that's causing exhaustive searches that are in vain.  Although in the cases where there are non-impossible routes, the well-intentioned ExplorerAI looking for the shortest route, and the well-intentioned Dijkstra short-circuiting for the fastest results assuming the start tile will change, happen to interact in a way that produces the slowest possible results.  But they're still dwarfed by the impossible galley searches.

My guess is the slow Warriors/Catapults/etc. may also be due to seeking impossible routes to water tiles.  They'll eventually get a path to somewhere they can reach, so they don't necessarily run into this every turn like the immobile galleys do.

Short term solution: Have the pathing return "no can do" for routes from land to sea (other than harbor to sea, if it's a sea unit).

Longer term, there's definitely the potential for caching if the algorithm detects the same start tile twice in a row, without the cache having been invalidated.  Or perhaps there's the possibility of adding a method that takes a list of tiles that we're looking for.  Either way I'm not surprised that we stumbled upon a worst-case combo with ExplorerAI; given what we know about trade routes and Civ3 turn times, I expected pathing optimization to be a big performance factor.  Just didn't expect it to crop up quite _this_ early.

**Comment by QuintillusCFC** [comment]
Plot twist - logging is slow.  Not sure why/what part of it is slow (beyond the Filter.ByIncludingOnly part of LogManager not making a difference).  But even if I just log a hard-coded string, in my seeded-random-number example where it looks at 6 paths, it takes the average from 2 to 11 milliseconds, or +1.5ms per path.  This means that the log statements are responsible for over 80% of the run time.

It's a sort of Heisenbug situation where applying debug methods to evaluate the state changes the characteristics of what you're trying to debug.

**Comment by QuintillusCFC** [comment]
Hmm, if I leave off the Write To Godot Sink, but leave on the log of a hard-coded string, the overhead from six logs shrinks from 9 ms to 1 ms.  That is a major reduction.

Gotcha: Leaving out that line also takes our the formatter.  So it could be the sink that's slow, it could be my console that's slow, or it could be the formatter that's slow.  But the log.Warning("hardcoded string") itself is not especially slow.

**Comment by QuintillusCFC** [comment]
Once I got the logs fixed up a bit, I figured out some of what's going on.  Basically, when a unit finished exploring an area, it needed a path to a new area to explore.  It evaluated _all_ the possible places it could explore, which took a long time, especially with fresh paths to the farthest places.  Now, it'll start with the closest tiles (as the crow flies), and if it can get to them quickly, it won't even bother with ones farther away.  In testing, this saved 85/86ths of the time in my Babylon test case.

This also allowed turns to be speedy until about turn 200, which is when they used to slow down.  While I'm not sure this is all of it, the improved logging also allowed me to see that every turn the Dutch AI was burning time asking for a route to tile 37, 27, which they knew existed but which was on another continent.  This led to an exhaustive search, taking 21 milliseconds, for a route for every request, and they were asking 155 times, for every spare unit, taking 3.255 seconds on turn 291.  They had over 40 cities by this point, so the number of units was increasing quickly.

The other AIs were also spazzing out a bit trying to send their units on boats (and some of the AI actually figured out how to move boats, I didn't know they could do that), but their actions did not take nearly as much time as the Dutch pathing requests.

I suspect there are other inefficiencies as well, which improved logs will help us spot.  Notably, the city graphics blink in the late game, and I'm sure they're not optimized given they were quite early on, so I suspect lots of city graphics = late game slowdowns as well.
